src/cpuid/src/cpuid_ffi.rs

```diff
@@ -0,0 +1,593 @@
+        if self.nent == 0 && n > 0 {
+            let new_layout = Layout::array::<RawKvmCpuidEntry>(n).unwrap();
+
+            // Ensure that the new allocation doesn't exceed `isize::MAX` bytes.
+            assert!(
+                isize::try_from(new_layout.size()).is_ok(),
+                "Allocation too large"
+            );
+
+            // SAFETY: Always safe.
+            let new_ptr = unsafe { std::alloc::alloc(new_layout) };
+            self.entries = match NonNull::new(new_ptr.cast::<RawKvmCpuidEntry>()) {
+                Some(p) => p,
+                None => std::alloc::handle_alloc_error(new_layout),
+            };
+        }
+        // realloc
+        else if self.nent > 0 && n > 0 {
+            let new_layout = Layout::array::<RawKvmCpuidEntry>(n).unwrap();
+
+            // Ensure that the new allocation doesn't exceed `isize::MAX` bytes.
+            assert!(
+                isize::try_from(new_layout.size()).is_ok(),
+                "Allocation too large"
+            );
+
+            let old_layout =
+                Layout::array::<RawKvmCpuidEntry>(usize::try_from(self.nent).unwrap()).unwrap();
+            let old_ptr = self.entries.as_ptr().cast::<u8>();
+            // SAFETY: Always safe.
+            let new_ptr = unsafe { std::alloc::realloc(old_ptr, old_layout, new_layout.size()) };
+
+            self.entries = match NonNull::new(new_ptr.cast::<RawKvmCpuidEntry>()) {
+                Some(p) => p,
+                None => std::alloc::handle_alloc_error(new_layout),
+            };
+        }
+        // dealloc
+        else if self.nent > 0 && n == 0 {
+            let old_layout =
+                Layout::array::<RawKvmCpuidEntry>(usize::try_from(self.nent).unwrap()).unwrap();
+            let old_ptr = self.entries.as_ptr().cast::<u8>();
+            // SAFETY: Always safe.
+            unsafe { std::alloc::dealloc(old_ptr, old_layout) };
+            self.entries = NonNull::dangling();
+        }
+        self.nent = u32::try_from(n).unwrap();
+    }
+
+    /// Pushes entry onto end.
+    ///
+    /// # Panics
+    ///
+    /// On allocation failure.
+    #[allow(
+        clippy::unwrap_used,
+        clippy::integer_arithmetic,
+        clippy::arithmetic_side_effects
+    )]
+    #[inline]
+    pub fn push(&mut self, entry: RawKvmCpuidEntry) {
+        self.resize(usize::try_from(self.nent).unwrap() + 1);
+        // SAFETY: Always safe.
+        unsafe {
+            std::ptr::write(
+                self.entries
+                    .as_ptr()
+                    .add(usize::try_from(self.nent).unwrap()),
+                entry,
+            );
+        }
+    }
+    /// Pops entry from end.
+    ///
+    /// # Panics
+    ///
+    /// On allocation failure.
+    #[allow(clippy::unwrap_used, clippy::unwrap_in_result)]
+    #[inline]
+    pub fn pop(&mut self) -> Option<RawKvmCpuidEntry> {
+        if self.nent > 0 {
+            let u_nent = usize::try_from(self.nent).unwrap();
```

Comment:
I've updated this function to be more safe and with additional documentation:
```rust
/// Pops entry from end.
///
/// # Panics
///
/// On allocation failure.
#[inline]
pub fn pop(&mut self) -> Option<RawKvmCpuidEntry> {
    (self.nent > 0).then(|| {
        // SAFETY: `usize` will always be at least 32 bits, thus `u32` can always be safely
        // converted into it.
        let len = unsafe { usize::try_from(self.nent).unwrap_unchecked() };
        // SAFETY: When `self.entries.as_ptr().add(u_nent)` contains a valid value.
        let rtn = unsafe { std::ptr::read(self.entries.as_ptr().add(len)) };

        // SAFETY: We check before `self.nent > 0` therefore unwrapping here is safe.
        let new_nent = unsafe { self.nent.checked_sub(1).unwrap_unchecked() };

        // Since we are decreasing the size `resize` should never panic.
        #[allow(clippy::unwrap_used)]
        self.resize(new_nent).unwrap();

        rtn
    })
}
```